src/Microsoft.VisualStudio.Threading/AsyncBarrier.cs

```diff
@@ -55,23 +62,31 @@ public Task SignalAndWait()
                 // Unleash everyone that preceded this one.
                 while (this.waiters.Count > 0)
                 {
-                    Task.Factory.StartNew(
-                        state => ((TaskCompletionSource<EmptyStruct>)state!).SetResult(default(EmptyStruct)),
-                        this.waiters.Pop(),
-                        CancellationToken.None,
-                        TaskCreationOptions.None,
-                        TaskScheduler.Default);
+                    this.waiters.Pop().TrySetResult(default);
                 }
 
                 // And allow this one to continue immediately.
-                return Task.CompletedTask;
+                return new ValueTask(cancellationToken.IsCancellationRequested
+                    ? Task.FromCanceled(cancellationToken)
+                    : Task.CompletedTask);
             }
             else
             {
                 // We need more folks. So suspend this caller.
-                var tcs = new TaskCompletionSource<EmptyStruct>();
+                TaskCompletionSource<EmptyStruct> tcs = new(TaskCreationOptions.RunContinuationsAsynchronously);
+#if NET
+                cancellationToken.Register(static (tcs, ct) => ((TaskCompletionSource<EmptyStruct>)tcs!).TrySetCanceled(ct), tcs);
+#else
+                cancellationToken.Register(
+                    static s =>
+                    {
+                        var t = (Tuple<TaskCompletionSource<EmptyStruct>, CancellationToken>)s!;
+                        t.Item1.TrySetCanceled(t.Item2);
+                    },
+                    Tuple.Create(tcs, cancellationToken));
+#endif
```

Comment:
💡 Only need to do this in the rare case of `CanBeCanceled`

Reply:
Nice. I'll restrict the check to the non-.NET code, since the .NET code doesn't allocate anything when it calls `Register`, and that method will no-op internally.